File: go/activation/softmax.go
package activation

import "github.com/flygerian/shapes"

func Softmax(ctx shapes.Context, logits shapes.Tensor) shapes.Tensor {
	ndims := uint32(len(logits.Shape()))
	classDim := ndims - 1

	maxLogits := logits.Max(ctx, classDim)
	shifted := logits.Minus(ctx, maxLogits)
	exp := shifted.Exp(ctx)
	probs := exp.Divide(ctx, exp.Sum(ctx, classDim))

	return probs
}

File: base/nn/softmax.c
#include "shapes.h"
#include "types.h"

Tensor *nn_Softmax(Context *ctx, Tensor *logits) {
  u8 ndims = logits->shape.numOfDims;
  u8 classDim = ndims - 1;

  Tensor *maxLogits = Max(ctx, logits, classDim);
  Tensor *shifted = Subtract(ctx, logits, maxLogits);
  Tensor *exp = Exp(ctx, shifted);
  Tensor *probs = Divide(ctx, exp, Sum(ctx, exp, classDim));

  return probs;
}
